File: features/times/src/main/java/com/metinkale/prayer/times/OngoingNotificationsReceiver.java
/*
 * Copyright (c) 2013-2019 Metin Kale
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.metinkale.prayer.times;

import android.app.Notification;
import android.app.NotificationManager;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.graphics.drawable.Icon;
import android.os.Build;
import android.os.SystemClock;
import android.text.Html;
import android.text.Spannable;
import android.text.style.StyleSpan;
import android.util.Pair;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RemoteViews;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.metinkale.prayer.Preferences;
import com.metinkale.prayer.receiver.InternalBroadcastReceiver;
import com.metinkale.prayer.service.ForegroundService;
import com.metinkale.prayer.times.fragments.TimesFragment;
import com.metinkale.prayer.times.times.Times;
import com.metinkale.prayer.times.times.Vakit;
import com.metinkale.prayer.times.utils.NotificationUtils;
import com.metinkale.prayer.utils.LocaleUtils;

import org.joda.time.DateTime;
import org.joda.time.LocalDate;
import org.joda.time.LocalDateTime;
import org.joda.time.LocalTime;
import org.joda.time.Period;
import org.joda.time.PeriodType;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class OngoingNotificationsReceiver extends InternalBroadcastReceiver implements InternalBroadcastReceiver.OnTimeTickListener, InternalBroadcastReceiver.OnPrefsChangedListener {
    private static final String FOREGROUND_NEEDY_ONGOING = "ongoing";
    private Integer mDefaultTextColor = null;

    @Override
    public void onTimeTick() {

        int textColor = Preferences.ONGOING_TEXT_COLOR.get();
        int bgColor = Preferences.ONGOING_BG_COLOR.get();

        NotificationManager notMan = (NotificationManager) getContext().getSystemService(Context.NOTIFICATION_SERVICE);

        LocalDate cal = LocalDate.now();

        List<Pair<Integer, Notification>> notifications = new ArrayList<>();
        for (Times t : Times.getTimes()) {
            if (!t.isOngoingNotificationActive()) {
                notMan.cancel(t.getIntID());
                continue;
            }


            boolean icon = Preferences.SHOW_ONGOING_ICON.get();
            boolean number = Preferences.SHOW_ONGOING_NUMBER.get();
            FirebaseCrashlytics.getInstance().setCustomKey("showIcon", icon);
            FirebaseCrashlytics.getInstance().setCustomKey("showNumber", number);

            RemoteViews views = new RemoteViews(getContext().getPackageName(), R.layout.notification_layout);
            if (Build.MANUFACTURER.toLowerCase(Locale.ENGLISH).contains("xiaomi")) {
                views.setViewPadding(R.id.notification, 0, 0, 0, 0);
            }

            if (bgColor != 0) {
                views.setInt(R.id.notification, "setBackgroundColor", bgColor);
            }
            views.setTextViewText(android.R.id.title, t.getName());

            if (textColor != 0)
                views.setTextColor(android.R.id.title, textColor);

            int[] timeIds = {R.id.time0, R.id.time1, R.id.time2, R.id.time3, R.id.time4, R.id.time5};
            int[] vakitIds = {R.id.fajr, R.id.sun, R.id.zuhr, R.id.asr, R.id.maghrib, R.id.ishaa};

            int marker = t.getCurrentTime();
            if (Preferences.VAKIT_INDICATOR_TYPE.get().equals("next")) {
                marker = marker + 1;
            }
            for (Vakit vakit : Vakit.values()) {
                LocalTime time = t.getTime(cal, vakit.ordinal()).toLocalTime();
                if (marker == vakit.ordinal()) {
                    views.setTextViewText(vakitIds[vakit.ordinal()], Html.fromHtml("<strong>" + vakit.getString() + "</strong>"));
                    if (Preferences.CLOCK_12H.get()) {
                        Spannable span = (Spannable) LocaleUtils.formatTimeForHTML(time);
                        span.setSpan(new StyleSpan(Typeface.BOLD), 0, span.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                        views.setTextViewText(timeIds[vakit.ordinal()], span);
                    } else {
                        views.setTextViewText(timeIds[vakit.ordinal()],
                                Html.fromHtml("<strong>" + LocaleUtils.formatTimeForHTML(time) + "</strong>"));
                    }
                } else {
                    views.setTextViewText(vakitIds[vakit.ordinal()], vakit.getString());
                    views.setTextViewText(timeIds[vakit.ordinal()], LocaleUtils.formatTimeForHTML(time));
                }
                if (textColor != 0) {
                    views.setTextColor(timeIds[vakit.ordinal()], textColor);
                    views.setTextColor(vakitIds[vakit.ordinal()], textColor);
                }
            }

            DateTime nextTime = t.getTime(cal, t.getNextTime()).toDateTime();
            if (Build.VERSION.SDK_INT >= 24 && Preferences.COUNTDOWN_TYPE.get().equals(Preferences.COUNTDOWN_TYPE_SHOW_SECONDS)) {
                views.setChronometer(R.id.countdown, nextTime.getMillis() - (System.currentTimeMillis() - SystemClock.elapsedRealtime()), null, true);
            } else {
                String txt = LocaleUtils.formatPeriod(DateTime.now(), nextTime, false);
                views.setString(R.id.countdown, "setFormat", txt);
                views.setChronometer(R.id.countdown, 0, txt, false);
            }
            if (textColor != 0) {
                views.setTextColor(R.id.countdown, textColor);
            }

            Notification.Builder builder =
                    new Notification.Builder(getContext());
            builder.setContentIntent(TimesFragment.getPendingIntent(t));

            if (!icon) {
                builder.setSmallIcon(R.drawable.ic_placeholder);
            } else if (number && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                builder.setSmallIcon(Icon.createWithBitmap(getIconFromMinutes(t)));
            } else {
                builder.setSmallIcon(R.drawable.ic_abicon);
            }
            builder.setOngoing(true);
            builder.setWhen(icon ? System.currentTimeMillis() : 0);


            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                builder.setCustomContentView(views);
            } else {
                builder.setContent(views);
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                builder.setChannelId(NotificationUtils.getOngoingChannel(getContext()));
            }


            Notification noti = builder.build();
            noti.priority = Notification.PRIORITY_LOW;
            notifications.add(new Pair<>(t.getIntID(), noti));
        }

        if (!notifications.isEmpty()) {
            for (int i = 0; i < notifications.size(); i++) {
                Pair<Integer, Notification> pair = notifications.get(i);
                if (i == 0 && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    ForegroundService.addNeedy(getContext(), FOREGROUND_NEEDY_ONGOING, pair.second, pair.first);
                } else {
                    notMan.notify(pair.first, pair.second);
                }
            }
        } else {
            ForegroundService.removeNeedy(getContext(), FOREGROUND_NEEDY_ONGOING);
        }

    }


    private Bitmap getIconFromMinutes(Times t) {
        int left = new Period(LocalDateTime.now(), t.getTime(LocalDate.now(), t.getNextTime()), PeriodType.minutes()).getMinutes();
        Resources r = getContext().getResources();

        int size = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 24, r.getDisplayMetrics());
        Bitmap b = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
        Canvas c = new Canvas(b);
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setColor(0xFFFFFFFF);
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setTextSize(size);
        paint.setTextSize(size * size / paint.measureText((left < 10 ? left * 10 : left) + ""));
        float yPos = c.getHeight() / 2f - (paint.descent() + paint.ascent()) / 2;
        c.drawText(left + "", size / 2f, yPos, paint);
        return b;
    }

    @Override
    public void onPrefsChanged(@NonNull String key) {
        if (key.equals(Preferences.SHOW_ONGOING_ICON.getKey())
                || key.equals(Preferences.SHOW_ONGOING_NUMBER.getKey())
                || key.equals(Preferences.ONGOING_TEXT_COLOR.getKey())
                || key.equals(Preferences.ONGOING_BG_COLOR.getKey())) {
            onTimeTick();
        }
    }


}

File: features/times/src/main/java/com/metinkale/prayer/times/OngoingNotificationsReceiver.kt
/*
 * Copyright (c) 2013-2023 Metin Kale
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.metinkale.prayer.times

import android.app.Notification
import android.app.NotificationManager
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface
import android.os.Build
import android.os.SystemClock
import android.text.Spannable
import android.text.style.StyleSpan
import android.util.Pair
import android.util.TypedValue
import android.view.View
import android.widget.RemoteViews
import androidx.core.app.NotificationCompat
import androidx.core.graphics.drawable.IconCompat
import androidx.core.text.toSpannable
import com.metinkale.prayer.CrashReporter.setCustomKey
import com.metinkale.prayer.Preferences
import com.metinkale.prayer.receiver.InternalBroadcastReceiver
import com.metinkale.prayer.receiver.InternalBroadcastReceiver.OnPrefsChangedListener
import com.metinkale.prayer.receiver.InternalBroadcastReceiver.OnTimeTickListener
import com.metinkale.prayer.service.ForegroundService
import com.metinkale.prayer.times.fragments.TimesFragment.Companion.getPendingIntent
import com.metinkale.prayer.times.times.*
import com.metinkale.prayer.times.utils.NotificationUtils
import com.metinkale.prayer.utils.LocaleUtils
import java.time.*

class OngoingNotificationsReceiver : InternalBroadcastReceiver(), OnTimeTickListener,
    OnPrefsChangedListener {
    private val textColor get() = Preferences.ONGOING_TEXT_COLOR.takeIf { it != 0 }
    private val bgColor get() = Preferences.ONGOING_BG_COLOR.takeIf { it != 0 }
    private val icon get() = Preferences.SHOW_ONGOING_ICON
    private val number get() = Preferences.SHOW_ONGOING_NUMBER

    override fun onTimeTick() {

        val notMan = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        val notifications: MutableList<Pair<Int, Notification>> = ArrayList()
        for (t in Times.current) {
            if (!t.ongoing) {
                notMan.cancel(t.id)
                continue
            }
            setCustomKey("showIcon", icon)
            setCustomKey("showNumber", number)


            val channelId = NotificationUtils.getOngoingChannel(context)
            val builder = NotificationCompat.Builder(context, channelId)
            builder.setContentIntent(getPendingIntent(t))
            if (!icon) {
                builder.setSmallIcon(R.drawable.ic_placeholder)
            } else if (number) {
                builder.setSmallIcon(IconCompat.createWithBitmap(getIconFromMinutes(t)))
            } else {
                builder.setSmallIcon(R.drawable.ic_abicon)
            }
            builder.setOngoing(true)
            builder.setWhen(if (icon) System.currentTimeMillis() else 0)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                builder.setStyle(NotificationCompat.DecoratedCustomViewStyle())
                builder.setCustomContentView(buildSmallRemoteView(t))
                builder.setCustomBigContentView(buildLargeRemoteView(t))
                builder.setShowWhen(false)
            } else {
                builder.setContent(buildLargeRemoteView(t))
            }
            val noti = builder.build()
            noti.priority = Notification.PRIORITY_LOW
            notifications.add(Pair(t.id, noti))
        }
        if (notifications.isNotEmpty()) {
            for (i in notifications.indices) {
                val pair = notifications[i]
                if (i == 0 && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    ForegroundService.addNeedy(
                        context,
                        FOREGROUND_NEEDY_ONGOING,
                        pair.second,
                        pair.first
                    )
                } else {
                    notMan.notify(pair.first, pair.second)
                }
            }
        } else {
            ForegroundService.removeNeedy(context, FOREGROUND_NEEDY_ONGOING)
        }
    }


    fun RemoteViews.initCommon(t: Times) = let { views ->

        val today = LocalDate.now()

        bgColor?.let { views.setInt(R.id.notification, "setBackgroundColor", it) }

        // Title
        views.setTextViewText(android.R.id.title, t.name)
        textColor?.let { views.setTextColor(android.R.id.title, it) }

        // Countdown
        val nextTime = t.getTime(today, t.getNextTime()).atZone(ZoneId.systemDefault()).toInstant()
        if (Build.VERSION.SDK_INT >= 24 && Preferences.COUNTDOWN_TYPE == Preferences.COUNTDOWN_TYPE_SHOW_SECONDS) {
            views.setChronometer(
                R.id.countdown,
                nextTime.toEpochMilli() - (System.currentTimeMillis() - SystemClock.elapsedRealtime()),
                null,
                true
            )
        } else {
            val txt = LocaleUtils.formatPeriod(Instant.now(), nextTime, false)
            views.setString(R.id.countdown, "setFormat", txt)
            views.setChronometer(R.id.countdown, 0, txt, false)
        }
        textColor?.let { views.setTextColor(R.id.countdown, it) }

    }

    private fun buildLargeRemoteView(t: Times): RemoteViews {
        val today = LocalDate.now()


        val nameIds = listOf(R.id.fajr, R.id.sun, R.id.zuhr, R.id.asr, R.id.maghrib, R.id.ishaa)
        val timeIds = listOf(R.id.time0, R.id.time1, R.id.time2, R.id.time3, R.id.time4, R.id.time5)


        val views = RemoteViews(context.packageName, R.layout.notification_layout_large)
        views.initCommon(t)

        // Times
        val marker = t.getCurrentTime()
            .let { if (Preferences.VAKIT_INDICATOR_TYPE == "next") it + 1 else it }
        for (vakit in Vakit.values()) {
            val time = t.getTime(today, vakit.ordinal).toLocalTime()

            fun Spannable.strongIfCurrent() = apply {
                if (marker == vakit.ordinal) {
                    setSpan(
                        StyleSpan(Typeface.BOLD),
                        0,
                        length,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
            }

            val timeString = LocaleUtils.formatTimeForHTML(time).toSpannable().strongIfCurrent()
            views.setTextViewText(timeIds[vakit.ordinal], timeString)

            val nameString = vakit.string.toSpannable().strongIfCurrent()
            views.setTextViewText(nameIds[vakit.ordinal], nameString)

            textColor?.let {
                views.setTextColor(timeIds[vakit.ordinal], it)
                views.setTextColor(nameIds[vakit.ordinal], it)
            }
        }




        return views
    }


    private fun buildSmallRemoteView(t: Times): RemoteViews {
        val views = RemoteViews(context.packageName, R.layout.notification_layout_small)
        views.initCommon(t)

        // Times
        views.setViewVisibility(R.id.times, View.GONE)
        views.setTextViewText(
            R.id.countdownprefix,
            Vakit.getByIndex(t.getCurrentTime()).string + " | "
        )
        return views
    }

    private fun getIconFromMinutes(t: Times): Bitmap {
        val left =
            Duration.between(LocalDateTime.now(), t.getTime(LocalDate.now(), t.getNextTime()))
                .toMinutes()
        val r = context.resources
        val size =
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 24f, r.displayMetrics).toInt()
        val b = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val c = Canvas(b)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG)
        paint.color = -0x1
        paint.textAlign = Paint.Align.CENTER
        paint.textSize = size.toFloat()
        paint.textSize =
            size * size / paint.measureText((if (left < 10) left * 10 else left).toString() + "")
        val yPos = c.height / 2f - (paint.descent() + paint.ascent()) / 2
        c.drawText(left.toString() + "", size / 2f, yPos, paint)
        return b
    }

    override fun onPrefsChanged(key: String) {
        if (key == "ongoingIcon" ||
            key == "ongoingNumber" ||
            key == "ongoingTextColor"
            || key == "ongoingBGColor"
        ) {
            onTimeTick()
        }
    }

    companion object {
        private const val FOREGROUND_NEEDY_ONGOING = "ongoing"
    }
}